src/librawspeed/common/Spline.h

```diff
@@ -0,0 +1,95 @@
+/*
+    RawSpeed - RAW file decoder.
+
+    Copyright (C) 2017 Robert Bieber
+
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+*/
+
+#pragma once
+
+#include "common/Common.h" // for ushort16
+#include "common/Point.h"  // for iPoint2D
+#include <vector>          // for vector
+
+namespace rawspeed {
+
+class Spline final {
+public:
+  static std::vector<ushort16>
+  calculateCurve(const std::vector<iPoint2D>& control_points) {
+    const int num_coords = control_points.size();
+    const int num_segments = num_coords - 1;
+
+    // These are the constant factors for each segment of the curve.
+    // Each segment i will have the formula:
+    // f(x) = a[i] + b[i]*(x - x[i]) + c[i]*(x - x[i])^2 + d[i]*(x - x[i])^3
+    std::vector<double> a(num_coords);
+    std::vector<double> b(num_segments);
+    std::vector<double> c(num_coords);
+    std::vector<double> d(num_segments);
+
+    // Extra values used during computation
+    std::vector<double> h(num_segments);
+    std::vector<double> alpha(num_segments);
+    std::vector<double> l(num_coords);
+    std::vector<double> mu(num_coords);
+    std::vector<double> z(num_coords);
+
+    for (int i = 0; i < num_coords; i++)
+      a[i] = control_points[i].y;
+
+    for (int i = 0; i < num_segments; i++)
+      h[i] = control_points[i + 1].x - control_points[i].x;
+
+    for (int i = 1; i < num_segments; i++)
+      alpha[i] =
+          (3. / h[i]) * (a[i + 1] - a[i]) - (3. / h[i - 1]) * (a[i] - a[i - 1]);
+
+    l[0] = mu[0] = z[0] = 0;
```

Comment:
@bieber but the https://en.wikipedia.org/wiki/Spline_(mathematics) says `l_0 = 1`?

Reply:
Ooh, yeah, I think I missed the 1 while reading through.  I'm not sure how consequential it is, but that's definitely a mistake

Reply:
I'll add some reference curves from [scipy.interpolate.CubicSpline](https://docs.scipy.org/doc/scipy-0.18.1/reference/generated/scipy.interpolate.CubicSpline.htm), and then see how this channges things.

Reply:
Actually, looking at the rest of the function I don't think l[0] is ever read (the next loop starts at 1 and references l[i] a bunch of times)